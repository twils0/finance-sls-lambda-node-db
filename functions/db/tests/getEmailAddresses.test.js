const { getEmailAddresses } = require('../getEmailAddresses');

global.pool = {
  query: jest.fn(),
};

const tickerCusip = 'testTickerCusip';
const securityId = 'testSecurityID';
const email1 = 'test1@test.com';
const emailAdditional1 = 'test_add1@test.com';
const email2 = 'test2@test.com';
const emailAddresses = [email1, emailAdditional1, email2];
const emptyResponse = {
  rows: [],
};
const securityReponse = {
  rows: [
    {
      security_id: securityId,
    },
  ],
};
const joinResponse = {
  rows: [
    {
      email: email1,
      email_additional: emailAdditional1,
    },
    {
      email: email2,
    },
  ],
};

describe('functions', () => {
  describe('db', () => {
    describe('getEmailAddresses', () => {
      afterEach(() => {
        global.pool.query.mockReset();
      });

      it('fails and returns unexpected error, when global.pool.query throws error first query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([tickerCusip]);
          return Promise.reject(error);
        });

        try {
          await getEmailAddresses(tickerCusip);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('fails and returns unexpected error, when security table query is empty', async () => {
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([tickerCusip]);
          return Promise.resolve(emptyResponse);
        });

        try {
          await getEmailAddresses(tickerCusip);
        } catch (errorCatch) {
          expect(errorCatch).toMatchSnapshot();
        }
      });

      it('fails and returns unexpected error, when global.pool.query throws error second query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([tickerCusip]);
          return Promise.resolve(securityReponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([securityId]);
          return Promise.reject(error);
        });

        try {
          await getEmailAddresses(tickerCusip);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('returns an empty array, when join table query is empty', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([tickerCusip]);
          return Promise.resolve(securityReponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([securityId]);
          return Promise.resolve(emptyResponse);
        });

        const result = await getEmailAddresses(tickerCusip);

        expect(result).toEqual({ emailAddresses: [] });
      });

      it('correctly returns email addresses', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([tickerCusip]);
          return Promise.resolve(securityReponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([securityId]);
          return Promise.resolve(joinResponse);
        });

        const result = await getEmailAddresses(tickerCusip);

        expect(result).toEqual({ emailAddresses });
      });
    });
  });
});
