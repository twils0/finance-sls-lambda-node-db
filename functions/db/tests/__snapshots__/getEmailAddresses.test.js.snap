// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions db getEmailAddresses correctly returns email addresses 1`] = `
"SELECT security_id
      FROM security.security
      WHERE ticker_cusip = $1::text"
`;

exports[`functions db getEmailAddresses correctly returns email addresses 2`] = `
"SELECT email, email_additional
        FROM \\"user\\".\\"user\\"
        WHERE user_id in (
        SELECT user_id
        FROM \\"user\\".join_user_security
        WHERE security_id = $1::uuid);"
`;

exports[`functions db getEmailAddresses fails and returns unexpected error, when global.pool.query throws error first query 1`] = `
"SELECT security_id
      FROM security.security
      WHERE ticker_cusip = $1::text"
`;

exports[`functions db getEmailAddresses fails and returns unexpected error, when global.pool.query throws error second query 1`] = `
"SELECT security_id
      FROM security.security
      WHERE ticker_cusip = $1::text"
`;

exports[`functions db getEmailAddresses fails and returns unexpected error, when global.pool.query throws error second query 2`] = `
"SELECT email, email_additional
        FROM \\"user\\".\\"user\\"
        WHERE user_id in (
        SELECT user_id
        FROM \\"user\\".join_user_security
        WHERE security_id = $1::uuid);"
`;

exports[`functions db getEmailAddresses fails and returns unexpected error, when security table query is empty 1`] = `
"SELECT security_id
      FROM security.security
      WHERE ticker_cusip = $1::text"
`;

exports[`functions db getEmailAddresses fails and returns unexpected error, when security table query is empty 2`] = `"unexpected error - ticker not found"`;

exports[`functions db getEmailAddresses returns an empty array, when join table query is empty 1`] = `
"SELECT security_id
      FROM security.security
      WHERE ticker_cusip = $1::text"
`;

exports[`functions db getEmailAddresses returns an empty array, when join table query is empty 2`] = `
"SELECT email, email_additional
        FROM \\"user\\".\\"user\\"
        WHERE user_id in (
        SELECT user_id
        FROM \\"user\\".join_user_security
        WHERE security_id = $1::uuid);"
`;
